server/handler.go

```diff
@@ -115,27 +142,34 @@ func (h *Handler) ComQuery(
 
 			r = nil
 			proccesedAtLeastOneBatch = true
-
 			continue
 		}
 
-		row, err := rows.Next()
-		if err != nil {
+		select {
+		case err := <-errchan:
 			if err == io.EOF {
-				break
+				goto EndFor
 			}
 
 			return err
-		}
+		case row := <-rowchan:
+			outputRow, err := rowToSQL(schema, row)
+			if err != nil {
+				return err
+			}
 
-		outputRow, err := rowToSQL(schema, row)
-		if err != nil {
-			return err
+			r.Rows = append(r.Rows, outputRow)
+			r.RowsAffected++
+		case <-time.After(waitTime):
```

Comment:
Since the default is quite long, and the goroutine on the timer will not exit till the timer fires or is cancelled, you should probably consider cancelling this explicitly:

```go
timeout := time.NewTimer(waitTime)
defer timeout.Stop()
select {
…
case <-timeout.C:
```

Reply:
I don't get you here, it doesn't matter if its long because it'll be cancelled as soon as there is another row or an error (or the timeout is exceeded). Or even if the `select` finish in one of the two other cases there will still be a hidden timer preventing the goroutine from being GC'ed?

Reply:
PS: To make my explanation clearer, I could use a much smaller waitTime (30 secs) and an accumulator for the total time to compare to the timeout to make it clearer that the wait in the `After` doesn't really matter (it will just loop over again until the timeout).

Reply:
However, good point about the loop: Since it's a loop we can't safely defer (which doesn't happen till the function exits), but instead `timeout.Stop()` after the end of the select.

Reply:
Wow I didn't knew time.After would create a goroutine. We definitely don't what that, even short duration ones, because some queries could iterateoover thousands of rows pretty quickly, so I'll switch to a Timer.

Reply:
If you really only need one of them for the whole scan—rather than one timeout per row—you could move the timer creation outside the loop and re-use the channel inside the `select`, that way you get only 1 additional goroutine for the whole operation (and of course reusing the unexpired timer is fine, as you mentioned).

Reply:
Optimization done! Let's reserve that RAM for Chrome :)